Error "ValueError: list.remove(x): x not in list" in testing on rolling

See [this log](https://github.com/felixdivo/ros2-easy-test/actions/runs/9572729155/job/26392699618?pr=44#step:11:121):

```txt
____________________ TestSingleNode.test_calling_an_action _____________________
ros2_easy_test/ros2_easy_test/decorators.py:128: in wrapper
    executor.spin_until_future_complete(
/opt/ros/rolling/lib/python3.12/site-packages/rclpy/executors.py:328: in spin_until_future_complete
    self.spin_once_until_future_complete(future, timeout_left)
/opt/ros/rolling/lib/python3.12/site-packages/rclpy/executors.py:907: in spin_once_until_future_complete
    self._spin_once_impl(timeout_sec, future.done)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <rclpy.executors.MultiThreadedExecutor object at 0x7fb80c2f32f0>
timeout_sec = <rclpy.executors.TimeoutObject object at 0x7fb80c41cb90>
wait_condition = <bound method Future.done of <rclpy.task.Task object at 0x7fb80c2f30b0>>

    def _spin_once_impl(
        self,
        timeout_sec: Optional[Union[float, TimeoutObject]] = None,
        wait_condition: Callable[[], bool] = lambda: False
    ) -> None:
        try:
            handler, entity, node = self.wait_for_ready_callbacks(
                timeout_sec, None, wait_condition)
        except ExternalShutdownException:
            pass
        except ShutdownException:
            pass
        except TimeoutException:
            pass
        except ConditionReachedException:
            pass
        else:
            self._executor.submit(handler)
            self._futures.append(handler)
            # make a copy of the list that we iterate over while modifying it
            # (https://stackoverflow.com/q/1207406/3753684)
            for future in self._futures[:]:
                if future.done():
>                   self._futures.remove(future)
E                   ValueError: list.remove(x): x not in list

/opt/ros/rolling/lib/python3.12/site-packages/rclpy/executors.py:895: ValueError
----------------------------- Captured stderr call -----------------------------
[INFO] [1718747826.149957309] [minimal_action_server]: Accepting goal request
[INFO] [1718747826.156829616] [minimal_action_server]: Executing goal...
[INFO] [1718747826.257690692] [minimal_action_server]: Publishing feedback: array('i', [0, 1, 1])
[INFO] [1718747826.268627448] [minimal_action_server]: Publishing feedback: array('i', [0, 1, 1, 2])
[INFO] [1718747826.279537183] [minimal_action_server]: Publishing feedback: array('i', [0, 1, 1, 2, 3])
[INFO] [1718747826.292523641] [minimal_action_server]: Returning result: array('i', [0, 1, 1, 2, 3])
[INFO] [1718747826.298843446] [minimal_action_server]: Accepting goal request
```

It appears to happen around [here](https://github.com/felixdivo/ros2-easy-test/blob/347a723242a7fdc6693706647b87c68882afabd2/ros2_easy_test/tests/test_interactions.py#L115-L117).

--- a/Dockerfile
+++ b/Dockerfile
@@ -15,7 +15,8 @@ RUN rosdep init && rosdep update
 ENV PIP_BREAK_SYSTEM_PACKAGES=1
 
 # Need to have setuptools version 64+ for editable installs
-RUN pip install --upgrade pip setuptools
+# Also fixes: https://github.com/pypa/setuptools/issues/4483#issuecomment-2236339726
+RUN pip install --upgrade pip "setuptools>=64.0" "packaging>=22.0"
 
 # Ensure sourced ROS environment at startup
 RUN echo 'source /opt/ros/$ROS_DISTRO/setup.bash' >> ~/.bashrc

--- a/ros2_easy_test/ros2_easy_test/decorators.py
+++ b/ros2_easy_test/ros2_easy_test/decorators.py
@@ -315,9 +315,9 @@ def with_launch_file(  # noqa: C901
                             node.destroy_node()
 
                         has_finished = executor.shutdown(shutdown_timeout)
-                        assert (
-                            has_finished
-                        ), f"Executor shutdown did not complete in {shutdown_timeout} seconds."
+                        assert has_finished, (
+                            f"Executor shutdown did not complete in {shutdown_timeout} seconds."
+                        )
 
                     # Make sure that the executor and the node are cleaned up/freed afterwards.
                     # Cleanup is critical for correctness since subsequent tests may NEVER reference old

--- a/ros2_easy_test/ros2_easy_test/env.py
+++ b/ros2_easy_test/ros2_easy_test/env.py
@@ -315,6 +315,8 @@ class ROS2TestEnvironment(Node):
             The result of the future
         """
 
+        # The event trick follows rclpy's service client implementation of the synchronous call()
+        # https://github.com/ros2/rclpy/blob/2b38e662b3635fffe5b68309f75e39f9e7441602/rclpy/rclpy/client.py#L97-L118
         event = Event()
 
         def unblock(_: Future) -> None:

--- a/ros2_easy_test/ros2_easy_test/env.py
+++ b/ros2_easy_test/ros2_easy_test/env.py
@@ -4,11 +4,11 @@
 from collections import defaultdict
 from importlib import import_module
 from queue import Empty, SimpleQueue
-from threading import RLock
+from threading import Event, RLock
 from time import monotonic, sleep
 
 # Typing
-from typing import Any, Dict, List, Mapping, Optional, Tuple, Type, cast
+from typing import Any, Dict, List, Mapping, Optional, Tuple, Type, TypeVar, cast
 
 # ROS
 from action_msgs.msg import GoalStatus
@@ -27,6 +27,9 @@ __all__ = ["ROS2TestEnvironment"]
 
 #: The normal timeout for asserts, like waiting for messages. This has to be surprisingly high.
 _DEFAULT_TIMEOUT: Optional[float] = 2
+
+
+AwaitedType = TypeVar("AwaitedType")
 
 
 class ROS2TestEnvironment(Node):
@@ -123,9 +126,9 @@ class ROS2TestEnvironment(Node):
         """
 
         with self._subscriber_mailboxes_lock:
-            assert (
-                topic in self._subscriber_mailboxes
-            ), f"topic {topic} it not being watched: please specify it in the constructor"
+            assert topic in self._subscriber_mailboxes, (
+                f"topic {topic} it not being watched: please specify it in the constructor"
+            )
 
             return self._subscriber_mailboxes[topic]
 
@@ -169,7 +172,7 @@ class ROS2TestEnvironment(Node):
             pass  # this is what we expect
         else:
             raise AssertionError(
-                f"A message was published on topic {topic} although none was expected: " f"{repr(message)}"
+                f"A message was published on topic {topic} although none was expected: {repr(message)}"
             ) from None
 
     def assert_message_published(self, topic: str, timeout: Optional[float] = _DEFAULT_TIMEOUT) -> RosMessage:
@@ -300,11 +303,12 @@ class ROS2TestEnvironment(Node):
             # There is not clear() in SimpleQueue
             self.listen_for_messages(topic, time_span=None)  # ignore the result
 
-    def await_future(self, future: Future, timeout: Optional[float] = 10) -> Any:
+    # In Rolling: Future[AwaitedType]
+    def await_future(self, future: Future, timeout: Optional[float] = 10) -> AwaitedType:
         """Waits for the given future to complete.
 
         Args:
-            future: The future to wait for
+            future: The ROS Future to wait for
             timeout: The timeout to wait for the future
 
         Raises:
@@ -315,15 +319,26 @@ class ROS2TestEnvironment(Node):
             The result of the future
         """
 
-        # This does not work with the default executor, so we use the one from the node
-        assert self.executor, "executor is not set"
-        # The type ignore is needed due to a bug in ROS2 Humble+
-        self.executor.spin_until_future_complete(future, timeout_sec=timeout)  # type: ignore[arg-type]
+        event = Event()
 
-        if future.done():
-            return future.result()
-        else:
-            raise TimeoutError(f"Future did not complete within {timeout} seconds")
+        def unblock(_: Future) -> None:
+            nonlocal event
+            event.set()
+
+        # This immediately calls the callback if the future is already done
+        future.add_done_callback(unblock)
+
+        # Check future.done() before waiting on the event.
+        # The callback might have been added after the future is completed,
+        # resulting in the event never being set.
+        if not future.done():
+            if not event.wait(timeout):
+                # Timed out. remove_pending_request() to free resources
+                self.remove_pending_request(future)
+                raise TimeoutError(f"Future did not complete within {timeout} seconds")
+
+        # Potential exception is raised here
+        return future.result()
 
     def _get_service_client(self, name: str, request_class: Type) -> Client:
         """Returns the service client for the given service name."""

--- a/ros2_easy_test/ros2_easy_test/env.py
+++ b/ros2_easy_test/ros2_easy_test/env.py
@@ -8,7 +8,7 @@ from threading import Event, RLock
 from time import monotonic, sleep
 
 # Typing
-from typing import Any, Dict, List, Mapping, Optional, Tuple, Type, TypeVar, cast
+from typing import Any, Dict, List, Mapping, Optional, Tuple, Type, cast
 
 # ROS
 from action_msgs.msg import GoalStatus
@@ -27,9 +27,6 @@ __all__ = ["ROS2TestEnvironment"]
 
 #: The normal timeout for asserts, like waiting for messages. This has to be surprisingly high.
 _DEFAULT_TIMEOUT: Optional[float] = 2
-
-
-AwaitedType = TypeVar("AwaitedType")
 
 
 class ROS2TestEnvironment(Node):
@@ -300,11 +297,10 @@ class ROS2TestEnvironment(Node):
                     self.clear_messages(topic=topic)
 
         else:
-            # There is not clear() in SimpleQueue
+            # There is no clear() in SimpleQueue
             self.listen_for_messages(topic, time_span=None)  # ignore the result
 
-    # In Rolling: Future[AwaitedType]
-    def await_future(self, future: Future, timeout: Optional[float] = 10) -> AwaitedType:
+    def await_future(self, future: Future, timeout: Optional[float] = 10) -> Any:
         """Waits for the given future to complete.
 
         Args:
@@ -328,14 +324,10 @@ class ROS2TestEnvironment(Node):
         # This immediately calls the callback if the future is already done
         future.add_done_callback(unblock)
 
-        # Check future.done() before waiting on the event.
-        # The callback might have been added after the future is completed,
-        # resulting in the event never being set.
-        if not future.done():
-            if not event.wait(timeout):
-                # Timed out. remove_pending_request() to free resources
-                self.remove_pending_request(future)
-                raise TimeoutError(f"Future did not complete within {timeout} seconds")
+        if not event.wait(timeout):
+            # Timed out. remove_pending_request() to free resources
+            self.remove_pending_request(future)
+            raise TimeoutError(f"Future did not complete within {timeout} seconds")
 
         # Potential exception is raised here
         return future.result()

